### PR TITLE
Allow appdata being stored in a separate object storage multibucket

### PR DESCRIPTION
**Pull request by MorrisJobke**
Disclaimer: This originated in a custom patchset and I just brought it into a bit better shape. It is still far from a proper solution, but I want to bring this up for discussions.

Why?

By default on a multibucket object storage setup the appdata files are all put into the bucket 0. This might cause some problems on various systems, where there is a lot of data (which are obviously the ones with an multibucket object store). The idea here is to allow to also put the appdata into a separate list of object store buckets to not have a lot of data (previews for example) in a single bucket.

What this does:

It basically replaces the object store instance on the fly (right before the write/read/unlink) operation with the one that fits the current path. It's definitely not a good way of handling this. Best would be to put this one layer above and register it as a proper Nextcloud storage.

Existing issues:

* `copyFromStorage` does only check the storage ID and it is the same as this is only a transparent replacement and thus it most likely will fail when something is moved between appdata buckets or out of appdata bucket into another one.
* `getObjectStore` is not covered and might randomly return one of the both instances (depending on what path was written/read from before)

### CODE REVIEW

**Comment by icewind1991** [comment]
I would much prefer mounting a separate `AppdataObjectStore` at the appdata folder than jamming the logic in here.

File: lib/private/Files/ObjectStore/ObjectStoreStorage.php [hunk 1]
```diff
@@ -591,6 +625,7 @@ private function copyFile(ICacheEntry $sourceEntry, string $to) {
 		$targetUrn = $this->getURN($targetEntry->getId());
 
 		try {
+			$this->verifyStorage($path);
```
Comment (nickvergessen) [inline lib/private/Files/ObjectStore/ObjectStoreStorage.php:628]:
```suggestion
			$this->verifyStorage($to);
```

**Comment by blizzz** [comment]
what's the state here?

**Comment by MorrisJobke** [comment]
> what's the state here?

Most likely this needs to be solved in a different way:



> I would much prefer mounting a separate `AppdataObjectStore` at the appdata folder than jamming the logic in here.


----

Let's close this one then.

**Pull request closed without merging**